java/src/jmri/jmrix/openlcb/swing/lccpro/LccProTableModel.java

```diff
@@ -0,0 +1,233 @@
+                } else {
+                    return null;
+                }
+            case UPGRADECOL:
+                if (pip.hasProtocol(ProtocolIdentification.Protocol.FirmwareUpgrade)) {
+                    return Bundle.getMessage("FieldUpgrade");
+                } else {
+                    return null;
+                }
+            default:
+                return "<unexpected column number>";
+        }
+    }
+
+    @Override
+    public void setValueAt(Object value, int row, int col) {
+        log.trace("getValue({}, {})", row, col);
+        MimicNodeStore.NodeMemo nodememo = nodestore.getNodeMemos().toArray(new MimicNodeStore.NodeMemo[0])[row];
+        if (nodememo == null) {
+            log.error("Button pushed but no corresponding node for row {}", row);
+            return;
+        }
+        var pip = nodememo.getProtocolIdentification();
+
+        switch (col) {
+            case CONFIGURECOL:
+                if (pip.hasProtocol(ProtocolIdentification.Protocol.ConfigurationDescription)) {
+                    var actions = new jmri.jmrix.openlcb.swing.ClientActions(memo.get(org.openlcb.OlcbInterface.class), memo);
+                    var node = nodememo.getNodeID();
+                    var description = jmri.jmrix.openlcb.swing.networktree.NetworkTreePane.augmentedNodeName(nodememo);
+                    actions.openCdiWindow(node, description);
+                }
+                break;
+            case UPGRADECOL:
+                if (pip.hasProtocol(ProtocolIdentification.Protocol.FirmwareUpgrade)) {
+                    var node = nodememo.getNodeID();
+                    var action = new jmri.jmrix.openlcb.swing.downloader.LoaderPane.Default(node);
+                    action.actionPerformed(null);
+                }
+                break;
+            default:
+                // TODO - fire the buttons
+                break;
+        }
+    }
+
+    public int getPreferredWidth(int column) {
+        int retval = 20; // always take some width
+        retval = Math.max(retval, new JLabel(getColumnName(column))
+            .getPreferredSize().width + 15);  // leave room for sorter arrow
+        for (int row = 0; row < getRowCount(); row++) {
+            if (getColumnClass(column).equals(String.class)) {
+                retval = Math.max(retval, new JLabel(getValueAt(row, column).toString()).getPreferredSize().width);
+            } else if (getColumnClass(column).equals(Integer.class)) {
+                retval = Math.max(retval, new JLabel(getValueAt(row, column).toString()).getPreferredSize().width);
+            } else if (getColumnClass(column).equals(ImageIcon.class)) {
+                retval = Math.max(retval, new JLabel((Icon) getValueAt(row, column)).getPreferredSize().width);
+            }
+        }
+        return retval + 5;
+    }
+
+
+    // drop listeners
+    public void dispose() {
+        // TODO - drop the listener for node changes
+    }
```

Comment:
It would be good to fix this method now.Otherwise it's easy to forget to do it later. It seems to me that the only thing that needs to be added is:

nodestore.removePropertyChangeListener(this);

Reply:
Done